docs: add local environment setup instructions

agraebe opened the pull request:
fixing #639

agraebe commented on src/pages/smart-contracts/setting-up-local-environment.md:
````diff
@@ -0,0 +1,332 @@
+The Stacks Explorer makes it easy way to review transactions. Let's spin up a local version:
+
+```bash
+git clone https://github.com/blockstack/explorer.git
````
@wileyj could we also just use a docker image for this project?

wileyj replied:
yes! we have an image that we use for the prod explorer. working on a fully-dockerized instruction set. 
the image is here: `quay.io/blockstack/explorer:latest`

agraebe commented on src/pages/smart-contracts/setting-up-local-environment.md:
````diff
@@ -0,0 +1,332 @@
+
+```bash
+cd ..
+git clone https://github.com/blockstack/ux.git
````
@wileyj here's another thing we could discuss: maybe this could be another docker image that we pull into the compose file for this tutorial?

wileyj replied:
For this I'll have to take a closer look, i'm not familiar with what's going on in this repo

wileyj replied:
update: totally doable. i'll have to submit a PR to get it ready but we can adjust instructions later once that is ready as part of the full-docker setup

agraebe commented:
@yknl could you help me identify the changes needed after monorepo work is done?

- [ ] use `StacksMocknet`
- [ ] change dependencies

hstove commented:
At a high level, we should use the testnet demo app for a "template" of clarity web app development. This way, you don't have to fiddle with React code. https://github.com/blockstack/ux/tree/master/packages/test-app

yknl commented on src/pages/smart-contracts/setting-up-local-environment.md:
```diff
@@ -0,0 +1,332 @@
+
+// set to local mocknet
+const network = new StacksTestnet();
+network.coreApiUrl = 'http://localhost:3999';
```
This can be changed to `const network = new StacksMocknet();` as soon as stacks.js monorepo is released and dependency updated.

friedger commented:
In my practise it is much easier to restart mocknet and redeploy the contract than finding the code for the contract addresses and replace them. In particular if you have more than one. 

The `yarn lerna.. ` command for the authenticator should be simplified to something like `yarn start-authenticator`

Could the explorer be part of stacks api docker image? Or have a mocknet docker image?

agraebe commented:
@dantrevino and @psq - any input for this? what is your local contract setup like?

dantrevino commented:
@agraebe ty for reaching out.  Instructions work and seem clear enough, however ... I feel like this is useful for either a) those getting started from scratch, and/or b) bigger teams that have a real distinction between their QA and dev teams.  Either that or I personally lack a good process, which is highly likely.

One nitpick ... align on a single node package manager. Instructions are mixed between npm and yarn.

None of this precludes the PRs usefulness.

For me personally, I have the following issues:
1) Setting up my own authenticator means I have to track a completely different set of different user accounts in a different set of browsers.  Probably good practice, but not very likely for lone developers.  I.e. I see a lot of value in testing with my real-world (test) accounts that I use across multiple apps, including apps that are not mine.  This could become even more required if Collections ever becomes reality.
2) Instructions for doing an `npm init clarity-starter` within an **existing** project would likely be easier for devs who want to bring Clarity to their existing applications.  It's not hard to do manually but would lower friction for their entry to Stacks, imho.

psq commented:
@agraebe my setup looks different, and probably more involved, but allows for running up to date, or even patched components.
- postgres runs natively, and I manually clear/recreate database to reset
- use the docker image to run bitcoind (the only thing that really never needs updating
- run stacks-node (in mining mocknet mode) from git with suitable branch (docker images are not updated fast enough for me)
- run stacks-node-api from git from suitable branch if needed
(so that's 3 terminal windows, with full access to the logs, and easy to restart)
I set all the ports explicitly, either via environment variable or configuration rather than rely on having the startup script to pick a new one each time (try that a 100 times and let me know how if felt 😂)
I some times have more than one environment running, so again, all ports need to be set explicitly, no guessing.

I usually don't run the explorer, while a great tools when things work, it does not expose enough to diagnose problems when there are errors, so I rely on logs.

To deploy processes, I rely on my own javascript code, when working on swapr+flexr, I need to redeploy 10 contracts, in the right order (https://github.com/psq/flexr/blob/master/deploy-contracts.ts#L86-L95)

I would rather see explanations on how you can change the auth url and api url in your app code rather than rely on pseudo magic code that makes the choice for you.  In practice, it rarely does pick the ***right*** ones and explaining how to use explicit values would save people some time when it makes the wrong choice.  Plus if you have to run on a server with your custom node (auth or api), learning how to set the url explicitly is something you have to do anyway.

And I sometimes have to run with my own authenticator, so that's one more dimension to add.

Also, I would recommend people start working on their clarity contracts using the js sdk, not even mocknet, it is a lot easier to get something up and running this way (example here https://github.com/psq/flexr/blob/master/test/unit/flexr.ts, @friedger probably has simpler examples)

friedger commented:
@agraebe , like psq I am not using testnet explorer locally.

For mocknet, 
* I am running the docket image of stacks-blockchain-api
* I run a script to deploy my contracts and restart the docker image whenever I want to retest
* I use one or more of the four test accounts that have initial balances (with a flag in the script to switch between test accounts and my "real" testnet accounts)

@psq With the recent update of the authenticator, a local instance is not necessary anymore, see #834

psq commented:
> @psq With the recent update of the authenticator, a local instance is not necessary anymore, see #834

yes, I noticed, but that was necessary earlier

agraebe commented:
**Wow, thanks everyone for the input!** I'm updating the PR and wanted to respond to some comments

--------

> This way, you don't have to fiddle with React code. https://github.com/blockstack/ux/tree/master/packages/test-app

@hstove I looked at it and figured it would be more difficult for folks (can't resolve multiple packages, need to clone the whole ux repo, not documented as well as the other app). Sticking to an app that has a tutorial is probably best for onboarding purposes

> In my practise it is much easier to restart mocknet and redeploy the contract than finding the code for the contract addresses and replace them. In particular if you have more than one.

@friedger Good point! I added instructions for that. Instead of the script, one can restart the mocknet using docker

> The yarn lerna.. command for the authenticator should be simplified to something like yarn start-authenticator

I removed the local authenticator setup now that the network can be set

> Could the explorer be part of stacks api docker image? Or have a mocknet docker image?

Yes! Thanks to @wileyj, the [docker compose project](https://github.com/blockstack/stacks-local-dev) includes the node+API+explorer

> One nitpick ... align on a single node package manager. Instructions are mixed between npm and yarn.

Thanks @dantrevino, I fixed that!

> Setting up my own authenticator means I have to track a completely different set of different user accounts in a different set of browsers

This is not required anymore, so we can avoid that complexity

> Instructions for doing an npm init clarity-starter within an existing project would likely be easier for devs who want to bring Clarity to their existing applications.

That's true! It is possible to just run the command inside a project folder, too. However, in my instructions I can basically say that one should complete the counter tutorial before doing this one. So, I can skip over the contract creation and focus on the goal of this tutorial: setup for development.

> I usually don't run the explorer, while a great tools when things work, it does not expose enough to diagnose problems when there are errors, so I rely on logs.

@psq and @friedger Yeah, I recognized that and removed the Explorer step/requirement. It is served through the docker compose, but I don't think I explicitly need to tell devs to set it up... also, thanks for sharing your local setup details! I believe it's probably too much for the regular smart contract dev, but still good for my understanding!

> I would rather see explanations on how you can change the auth url and api url in your app code rather than rely on pseudo magic code that makes the choice for you

I updated the instructions to cover that

> Also, I would recommend people start working on their clarity contracts using the js sdk, not even mocknet, it is a lot easier to get something up and running this way

Agreed! This tutorial requires the counter tutorial to be completed, which in fact uses the clarity project generator that comes with the js sdk :) So devs can work in TDD fashion before ever deploying to mocknet. I added that to the instructions too

> I use one or more of the four test accounts that have initial balances (with a flag in the script to switch between test accounts and my "real" testnet accounts)

@friedger thanks for sharing that! The new docker compose uses test accounts and I added instructions for deployment using one of them as well (with a note so folks know about it). The script seems nice! I'm not sure it's the right place to mention it but let me think about it

agraebe commented:
Current blockers:

- [ ] merging https://github.com/blockstack/stacks-local-dev/pull/2

CharlieC3 commented on src/pages/smart-contracts/setting-up-local-environment.md:
```diff
@@ -0,0 +1,206 @@
+
+~> The Stacks Node is running on port `20443`, the Stacks Blockchain API on `3999` and the Stacks Explorer is on port `3000`.
+
+## Step 3: Deploy your contract
```
@wileyj Whoops! Where's step 2?

agraebe replied:
thanks, fixed!

agraebe commented on src/pages/smart-contracts/setting-up-local-environment.md:
````diff
@@ -0,0 +1,332 @@
+yarn dev
+```
+
+You should now be able to open up `http://localhost:3000/` and see the homepage. **Make sure to select `Localhost` from the network list on the top right side of the screen!**
````
that;d be great! did we hash out the CORS concerns that I believe were brought up last time we chatted?

aulneau replied:
This will not be possible without a good amount of work. I believe the best answer is to have people run the explorer locally if they need to.

agraebe replied:
makes sense, thanks!

agraebe commented:
@yknl this can't be merged until https://github.com/blockstack/stacks.js/issues/878 is resolved. could you take a look please?

agraebe commented on src/pages/write-smart-contracts/setting-up-local-environment.md:
```diff
@@ -0,0 +1,206 @@
+
+# deploy the contract:
+# deploy_contract <SOURCE_FILE> <CONTRACT_NAME> <FEE> <NONCE> <PAYMENT_KEY> -t -T <NETWORK_ADDRESS>
+stx deploy_contract ./counter.clar counter 2000 0 cb3df38053d132895220b9ce471f6b676db5b9bf0b4adefb55f2118ece2478df01 -t -T http://localhost:20443
```
@CharlieC3 @wileyj I tried the explorer sandbox and the CLI to deploy to the mocknet but I never get any transaction 🤔 is this image maybe outdated/broken? are you able to broadcast a tx to a mocknet?

aulneau commented on src/pages/write-smart-contracts/setting-up-local-environment.md:
````diff
@@ -0,0 +1,201 @@
+docker-compose up -d
+```
+
+-> Make sure your Docker service is up and running
````
I'd also add a bit more content around docker and docker-compose

wileyj replied:
what would you suggest?

aulneau commented on src/pages/write-smart-contracts/setting-up-local-environment.md:
````diff
@@ -0,0 +1,201 @@
+{"status":"ready"}
+```
+
+~> The Stacks Node is running on port `20443`, the Stacks Blockchain API on `3999` and the Stacks Explorer is on port `3000`.
````
When I pulled the latest repo, the explorer is disabled by default. we should note that.

wileyj commented on src/pages/write-smart-contracts/setting-up-local-environment.md:
````diff
@@ -0,0 +1,201 @@
+```bash
+git clone https://github.com/blockstack/stacks-local-dev
+cd stacks-local-dev/
+docker-compose up -d
````
```
10:36:21 ~$ docker-compose up --help
...

Options:
    -d, --detach               Detached mode: Run containers in the background,
                               print new container names. Incompatible with
                               --abort-on-container-exit.
```

alexkeating commented on src/pages/write-smart-contracts/setting-up-local-environment.md:
```diff
@@ -0,0 +1,201 @@
+
+# deploy the contract:
+# deploy_contract <SOURCE_FILE> <CONTRACT_NAME> <FEE> <NONCE> <PAYMENT_KEY> -t -I <NETWORK_ADDRESS>
+stx deploy_contract ./counter.clar counter 2000 0 cb3df38053d132895220b9ce471f6b676db5b9bf0b4adefb55f2118ece2478df01 -t -I http://localhost:20443
```
Should this be -I or -T? I kept getting a NotEnoughFunds error until I found [this](https://github.com/blockstack/stacks.js/issues/840) issue and switched to -T.

agraebe replied:
adding @yknl for confirmation but in this case, I think it should be `-I`, right? There are 3 _kind of relevant_ options (`T`, `H`, and `I`) and I'm not certain about this one

alexkeating replied:
Okay, yea, found [this](https://github.com/blockstack/stacks.js/issues/878) and I think all my transactions are going to the testnet with -T.

agraebe commented:
this is quite outdated by now and there seems to be no strong demand for this type of tutorial right now. closing for now.
